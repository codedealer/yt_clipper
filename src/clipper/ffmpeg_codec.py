from fractions import Fraction
from typing import Optional, Tuple

from clipper.clipper_types import DictStrAny
from clipper.ytc_logger import logger


def getFfmpegVideoCodecArgs(
    videoCodec: str,
    cbr: Optional[int],
    mp: DictStrAny,
    mps: DictStrAny,
    qmax: int,
    qmin: int,
) -> Tuple[str, str, str]:
    if videoCodec in {"vp9", "vp8"}:
        return getFfmpegVideoCodecVpx(
            videoCodec=videoCodec,
            cbr=cbr,
            mp=mp,
            mps=mps,
            qmax=qmax,
            qmin=qmin,
        )

    if videoCodec == "h264":
        return getFfmpegVideoCodecH264(cbr=cbr, mp=mp, mps=mps, qmax=qmax, qmin=qmin)

    if videoCodec == "h264_vulkan":
        return getFfmpegVideoCodecH264Vulkan(cbr=cbr, mp=mp, mps=mps, qmax=qmax, qmin=qmin)

    if videoCodec == "h264_nvenc":
        return getFfmpegVideoCodecH264Nvenc(cbr=cbr, mp=mp, mps=mps, qmax=qmax, qmin=qmin)

    raise ValueError(f"Invalid video codec: {videoCodec}")


def getFfmpegVideoCodecVpx(
    videoCodec: str,
    cbr: Optional[int],
    mp: DictStrAny,
    mps: DictStrAny,
    qmax: int,
    qmin: int,
) -> Tuple[str, str, str]:
    if mps["minterpFPS"] is not None:
        fps_arg = f'-r {mps["minterpFPS"]}'
    elif not mp["isVariableSpeed"]:
        fps_arg = f'-r ({mps["r_frame_rate"]}*{mp["speed"]})'
    else:
        fps_arg = "-fps_mode vfr"

    sdr_args = "-pix_fmt yuv420p"
    hdr_args = "-profile:v 2 -pix_fmt yuv420p10le -color_primaries bt2020 -color_trc smpte2084 -colorspace bt2020nc"

    dynamic_range_args = sdr_args
    if mps["enableHDR"]:
        if videoCodec == "vp8":
            logger.warning("HDR output was requested but vp8 does not support HDR.")
        else:
            dynamic_range_args = hdr_args

    video_codec_args = " ".join(
        (
            f"-c:v libvpx-vp9" if videoCodec != "vp8" else f"-c:v libvpx",
            dynamic_range_args,
            f"-slices 8",
            f"-aq-mode 4 -row-mt 1 -tile-columns 6 -tile-rows 2" if videoCodec != "vp8" else "",
            f'-qmin {qmin} -crf {mps["crf"]} -qmax {qmax}' if mps["targetSize"] <= 0 else "",
            f'-b:v {mps["targetMaxBitrate"]}k' if cbr is None else f"-b:v {cbr}MB",
            f'-force_key_frames 1 -g {mp["averageSpeed"] * Fraction(mps["r_frame_rate"])}',
        ),
    )
    video_codec_input_args = ""
    video_codec_output_args = " ".join(("-f webm", fps_arg))
    return video_codec_args, video_codec_input_args, video_codec_output_args


def getFfmpegVideoCodecH264(
    cbr: Optional[int],
    mp: DictStrAny,
    mps: DictStrAny,
    qmax: int,
    qmin: int,
) -> Tuple[str, str, str]:
    """Following recommendations from https://www.lighterra.com/papers/videoencodingh264"""
    fps_arg = ""
    if not mps["h264DisableReduceStutter"]:
        if mps["minterpFPS"] is not None:
            fps_arg = f'-r {mps["minterpFPS"]}'
        elif not mp["isVariableSpeed"]:
            fps_arg = f'-r ({mps["r_frame_rate"]}*{mp["speed"]})'

    if mp["isVariableSpeed"]:
        fps_arg = "-fps_mode vfr"

    pixel_count = mps["width"] * mps["height"]

    # The me_method and me_range have a fairly significant impact on encoding speed and could be tweaked
    # Currently going with a high me_method (umh vs the default hex), and a moderate me_range (32 for higher resolutions vs the default 16)
    me_range = 16 if pixel_count < (1800 * 1000) else 32

    sdr_args = "-pix_fmt yuv420p"
    hdr_args = (
        "-pix_fmt yuv420p10le -color_primaries bt2020 -color_trc smpte2084 -colorspace bt2020nc"
    )

    dynamic_range_args = sdr_args
    if mps["enableHDR"]:
        dynamic_range_args = hdr_args

    video_codec_args = " ".join(
        (
            f"-c:v libx264",
            f"-movflags write_colr",
            dynamic_range_args,
            # h264 crf is between 0 and 51 (same as qp)
            f'-qmin 3 -crf {mps["crf"]} -qmax {qmax}' if mps["targetSize"] <= 0 else "",
            f'-b:v {mps["targetMaxBitrate"]}k' if cbr is None else f"-b:v {cbr}MB",
            f'-force_key_frames 1 -g {mp["averageSpeed"] * Fraction(mps["r_frame_rate"])}',
            # video_track_timescale = 2^4 * 3^2 * 5^2 * 7 * 11 * 13 * 23, max is ~2E9
            f" -video_track_timescale 82882800",
            "-bf 5",
            "-refs 4",
            "-qcomp 0.9",
            f"-aq-mode 4",
            "-rc-lookahead 40",
            "-weightb 1 -weightp 2",
            "-direct-pred auto",
            "-b-pyramid none",
            "-me_method umh",
            f"-me_range {me_range}",
            "-psy-rd 1.0:1.0",
            "-fastfirstpass 1",
            "-keyint_min 1",
            "-trellis 2",
            "-x264-params rc-lookahead=40",
        ),
    )

    video_codec_input_args = ""
    video_codec_output_args = " ".join(("-f mp4", fps_arg))
    return video_codec_args, video_codec_input_args, video_codec_output_args


def getFfmpegVideoCodecH264Vulkan(
    cbr: Optional[int],
    mp: DictStrAny,
    mps: DictStrAny,
    qmax: int,
    qmin: int,
) -> Tuple[str, str, str]:
    fps_arg = ""
    if not mps["h264DisableReduceStutter"]:
        if mps["minterpFPS"] is not None:
            fps_arg = f'-r {mps["minterpFPS"]}'
        elif not mp["isVariableSpeed"]:
            fps_arg = f'-r ({mps["r_frame_rate"]}*{mp["speed"]})'

    if mp["isVariableSpeed"]:
        fps_arg = "-fps_mode vfr"

    sdr_args = "-pix_fmt vulkan"
    # x2rgb10 may better support 10-bit HDR output with vulkan in the future
    # It doesn't seem to be allowed currently however, only the vulkan pixel format works with vulkan encoders
    # See https://patchwork.ffmpeg.org/project/ffmpeg/patch/1587532983-20287-1-git-send-email-fei.w.wang@intel.com/#55407
    hdr_args = "-pix_fmt vulkan -color_primaries bt2020 -color_trc smpte2084 -colorspace bt2020nc"

    dynamic_range_args = sdr_args
    if mps["enableHDR"]:
        dynamic_range_args = hdr_args
    video_codec_args = " ".join(
        (
            f"-c:v h264_vulkan",
            f"-movflags write_colr",
            dynamic_range_args,
            "-quality 0",
            "-rc_mode vbr",
            "-tune hq",
            # i_qfactor = QP factor between P and I frames,
            # b_qfactor = between P and B frames.
            "-i_qfactor 0.75 -b_qfactor 1.1",
            # doesn't support crf, qp ranges from -1 to 255
            f"-qmin 3 -qmax {qmax}" if mps["targetSize"] <= 0 else "",
            f'-b:v {mps["targetMaxBitrate"]}k' if cbr is None else f"-b:v {cbr}MB",
            f'-bufsize {mps["targetMaxBitrate"]}k' if cbr is None else f"-bufsize {cbr}MB",
            f'-maxrate {mps["targetMaxBitrate"]*4}k' if cbr is None else f"-maxrate {cbr*4}MB",
            f'-force_key_frames 1 -g {mp["averageSpeed"] * Fraction(mps["r_frame_rate"])}',
            # video_track_timescale = 2^4 * 3^2 * 5^2 * 7 * 11 * 13 * 23, max is ~2E9
            f" -video_track_timescale 82882800",
            "-qcomp 0.9",
            "-bf 5",
            "-refs 4",
            # x264 options that do not apply to h264_vulkan
            # f"-aq-mode 4",
            # "-weightb 1 -weightp 2",
            # "-direct-pred auto",
            # "-b-pyramid none",
            # "-me_method umh",
            # f"-me_range {me_range}",
            # "-psy-rd 1.0:1.0",
            # "-fastfirstpass 1",
            # "-x264-params rc-lookahead=40",
            "-keyint_min 1",
            "-trellis 2",
            # explicit defaults for h264_vulkan
            "-b_depth 1",
            "-async_depth 2",
        ),
    )

    video_codec_input_args = "-hwaccel vulkan -hwaccel_output_format vulkan"
    video_codec_output_args = " ".join(("-f mp4", fps_arg))
    return video_codec_args, video_codec_input_args, video_codec_output_args


def getFfmpegVideoCodecH264Nvenc(
    cbr: Optional[int],
    mp: DictStrAny,
    mps: DictStrAny,
    qmax: int,
    qmin: int,
) -> Tuple[str, str, str]:
    fps_arg = ""
    if not mps["h264DisableReduceStutter"]:
        if mps["minterpFPS"] is not None:
            fps_arg = f'-r {mps["minterpFPS"]}'
        elif not mp["isVariableSpeed"]:
            fps_arg = f'-r ({mps["r_frame_rate"]}*{mp["speed"]})'

    if mp["isVariableSpeed"]:
        fps_arg = "-fps_mode vfr"

    sdr_args = "-pix_fmt cuda"
    # h264_nvenc does not support 10-bit HDR output
    hdr_args = "-pix_fmt cuda -color_primaries bt2020 -color_trc smpte2084 -colorspace bt2020nc"

    dynamic_range_args = sdr_args
    if mps["enableHDR"]:
        dynamic_range_args = hdr_args

    video_codec_args = " ".join(
        (
            f"-c:v h264_nvenc",
            f"-movflags write_colr",
            dynamic_range_args,
            # vbr_hq is deprecated
            "-rc vbr" if cbr is None else "-rc cbr",
            # h264_nvenc cq should be equivalent to h264 crf and is between 0 and 51 (same as qp)
            # h264_nvenc doesn't support modes like constrained encoding / VBV (Video Buffering Verifier)
            # so we remove the -cq option when in cbr mode
            f"-cq {mps['crf']}" if cbr is None and mps["targetSize"] <= 0 else "",
            f"-qmin 3 -qmax {qmax}" if mps["targetSize"] <= 0 else "",
            f'-b:v {mps["targetMaxBitrate"]}k' if cbr is None else f"-b:v {cbr}MB",
            f'-maxrate {mps["targetMaxBitrate"]*2}k' if cbr is None else f"-maxrate {cbr*2}MB",
            f'-force_key_frames 1 -g {mp["averageSpeed"] * Fraction(mps["r_frame_rate"])}',
            # video_track_timescale = 2^4 * 3^2 * 5^2 * 7 * 11 * 13 * 23, max is ~2E9
            f" -video_track_timescale 82882800",
            "-qcomp 0.9",
            # NVENC specific settings - optimized for latency tolerant high quality encoding
            "-tune hq",
            "-preset p6",
            # There are no devices that support more than 4 B-frames currently
            "-bf 4",
            # rc-lookahead default is 40, diminishing returns beyond 60
            # although expensive, it may be worth it for gpu encodes which often have lower quality
            "-rc-lookahead 60",
            "-spatial-aq 1",
            "-temporal_aq 1",
            "-aq-strength 12",
            "-keyint_min 1",
            # weighted predictions are not supported with b-frames with nvenc
            # "-weighted_pred 1",
        ),
    )

    video_codec_input_args = "-hwaccel cuda -hwaccel_output_format cuda"
    video_codec_output_args = " ".join(("-f mp4", fps_arg))
    return video_codec_args, video_codec_input_args, video_codec_output_args
