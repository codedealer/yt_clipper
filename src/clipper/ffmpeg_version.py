from subprocess import PIPE, Popen

def getFfmpegVersion(ffmpeg_path: str) -> str:
    try:
        with Popen([ffmpeg_path, "-version"], stdout=PIPE, stderr=PIPE) as proc:
            if proc.stdout is not None:
                version_line = proc.stdout.readline()
            else:
                version_line = "unknown"
    except:  # pylint: disable=bare-except
        version_line = "unknown"

    return str(version_line).strip()
