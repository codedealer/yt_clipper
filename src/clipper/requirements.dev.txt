-r requirements.txt
pyinstaller == 5.0.1
