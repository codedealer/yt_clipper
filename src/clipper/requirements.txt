# Automatically generated by https://github.com/damnever/pigar.

coloredlogs == 14.0
notify_py == 0.2.3
verboselogs == 1.7
webvtt_py == 0.4.5
https://github.com/exwm/yt-dlc/archive/98a668862f6e8e4191557b272a7f2142d099a282.tar.gz#egg=youtube_dlc

