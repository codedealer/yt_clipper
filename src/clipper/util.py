import base64
import hashlib
import re
from importlib import util as importlib_util
from typing import Any, Dict, Union


def is_module_available(mod: str) -> bool:
    return importlib_util.find_spec(mod) is not None


def dictTryGetKeys(d: Dict, *keys: str, default=None) -> Any:
    for key in keys:
        value = d.get(key)
        if value is not None:
            return value

    return default


def notifyOnComplete(titleSuffix: str) -> None:
    from notifypy import Notify

    n = Notify()
    n.application_name = "yt_clipper"
    n.title = "yt_clipper Completed Run"
    n.message = f"Processed {titleSuffix}.json."
    n.send(block=False)


def getTrimmedBase64Hash(string: str, n_bytes: int = 9) -> str:
    hash_object = hashlib.sha256(string.encode(encoding="utf-8", errors="replace"))
    hex_dig = hash_object.digest()[:n_bytes]
    return base64.b64encode(hex_dig).decode("ascii")


def escapeSingleQuotesFFmpeg(string: str) -> str:
    return re.sub(r"'", r"'\\''", string)


def floorToEven(x: Union[int, str, float]) -> int:
    x = int(x)
    return x & ~1
