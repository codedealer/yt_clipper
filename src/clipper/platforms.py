import sys

from clipper.clipper_types import KnownPlatform, Settings
from clipper.ytc_logger import logger


def getFfmpegHeaders(platform: str) -> str:
    if platform == KnownPlatform.afreecatv.name:
        return " ".join(
            (
                f"-headers 'Referer: https://play.afreecatv.com/'",
                f"-headers 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:75.0) Gecko/20100101 Firefox/75.0'",
            ),
        )

    return ""


def getVideoPageURL(settings: Settings, platform: str, videoID: str) -> str:
    if platform == KnownPlatform.youtube.name:
        return f"https://www.youtube.com/watch?v={videoID}"
    if platform == KnownPlatform.vlive.name:
        return f"https://www.vlive.tv/video/{videoID}"
    if platform == KnownPlatform.naver_now_watch.name:
        return f"https://now.naver.com/watch/{videoID}"
    if platform == KnownPlatform.weverse.name:
        return settings["videoUrl"]
    if platform == KnownPlatform.naver_tv.name:
        return f"https://tv.naver.com/v/{videoID}"
    if platform == KnownPlatform.afreecatv.name:
        return settings["videoUrl"]

    logger.fatal(f"Unknown platform: {platform}")
    sys.exit(1)
