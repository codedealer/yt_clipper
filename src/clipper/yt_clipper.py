#!/usr/bin/env python3

import os
import sys
from pathlib import Path

import certifi

from clipper import (
    argparser,
    clip_maker,
    clipper_types,
    util,
    ytc_logger,
    ytc_settings,
)
from clipper.clipper_types import ClipperState
from clipper.ffmpeg_version import getFfmpegVersion
from clipper.version import __version__
from clipper.ytc_logger import logger
from clipper.ytdl import ytdl_bin_get_version

UNKNOWN_PROPERTY = "unknown"


def main() -> None:
    cs = clipper_types.ClipperState()
    setupDepPaths(cs)

    args, unknown, defArgs, argFiles = argparser.getArgs(cs.clipper_paths)

    cs.settings.update({"color_space": None, **args})
    ytc_settings.loadSettings(cs.settings)

    setupOutputPaths(cs)

    ytc_logger.setUpLogger(cs)

    logger.report(f"yt_clipper version: {__version__}")
    logger.report(
        f"yt-dlp version: {ytdl_bin_get_version(cs.clipper_paths)}",
    )
    logger.report(f"{getFfmpegVersion(cs.clipper_paths.ffmpegPath)}", extra={"highlighter": None})
    logger.info("-" * 80)

    logger.debug(f"The following  arguments were read from the command line {sys.argv[1:]}:")

    if defArgs:
        logger.notice(f"The following default arguments were read from {argFiles}:")
        logger.notice(defArgs)
        logger.info("-" * 80)
    elif argFiles:
        logger.notice(f"No uncommented arguments were found in {argFiles}")
        logger.info("-" * 80)

    if unknown:
        logger.notice(
            f"The following unknown arguments were provided and were ignored:",
        )
        logger.notice(unknown)
        logger.info("-" * 80)

    enableMinterpEnhancements(cs)

    ytc_settings.getInputVideo(cs)

    ytc_settings.getGlobalSettings(cs)

    logger.info("-" * 80)
    if not cs.settings["preview"]:
        clip_maker.makeClips(cs)
    else:
        clip_maker.previewClips(cs)

    ytc_logger.printReport(cs)

    if cs.settings["notifyOnCompletion"]:
        util.notifyOnComplete(cs.settings["titleSuffix"])


def setupDepPaths(cs: ClipperState) -> None:
    cp = cs.clipper_paths

    if getattr(sys, "frozen", False):
        cp.ffmpegPath = "./bin/ffmpeg"
        cp.ffprobePath = "./bin/ffprobe"
        cp.ffplayPath = "./bin/ffplay"
        cp.ytdlPath = "./bin/yt-dlp"
        if sys.platform == "win32":
            cp.ffmpegPath += ".exe"
            cp.ffprobePath += ".exe"
            cp.ffplayPath += ".exe"
            cp.ytdlPath += ".exe"

        if sys.platform == "darwin":
            cp.ytdlPath += "_macos"

            certifi_cacert_path = certifi.where()
            os.environ["SSL_CERT_FILE"] = certifi_cacert_path
            os.environ["REQUESTS_CA_BUNDLE"] = certifi_cacert_path


def setupOutputPaths(cs: ClipperState) -> None:
    settings = cs.settings
    cp = cs.clipper_paths
    cp.clipsPath += f'/{settings["titleSuffix"]}'

    os.makedirs(f"{cp.clipsPath}/temp", exist_ok=True)
    settings["downloadVideoPath"] = f'{cp.clipsPath}/{settings["downloadVideoNameStem"]}'


def enableMinterpEnhancements(cm: ClipperState) -> None:
    settings = cm.settings
    cp = cm.clipper_paths
    if settings["enableMinterpEnhancements"] and sys.platform == "win32":
        cp.ffmpegPath = "./bin/ffmpeg_ytc.exe"
        if not Path(cp.ffmpegPath).is_file():
            logger.critical(
                f"{cp.ffmpegPath} required for minterp enhancements not found.",
            )
            sys.exit(1)
        else:
            logger.success(f"Found {cp.ffmpegPath}. Minterp enhancements enabled.")
    else:
        settings["enableMinterpEnhancements"] = False


if __name__ == "__main__":
    main()
