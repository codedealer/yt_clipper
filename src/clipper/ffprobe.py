import json
import shlex
import subprocess
import time
from typing import Optional

from clipper.clipper_types import ClipperState, DictStrAny
from clipper.platforms import getFfmpegHeaders
from clipper.ytc_logger import logger


def ffprobeVideoProperties(cs: ClipperState, videoURL: str) -> Optional[DictStrAny]:
    cp = cs.clipper_paths
    settings = cs.settings

    ffprobeRetries = 3
    done = False
    while ffprobeRetries > 0 and not done:
        ffprobeRetries -= 1
        try:
            ffprobeFlags = " ".join(
                (
                    "-v quiet -select_streams v -print_format json -show_streams -show_format",
                    getFfmpegHeaders(settings["platform"]),
                ),
            )
            ffprobeCommand = f'"{cp.ffprobePath}" "{videoURL}" {ffprobeFlags} '
            ffprobeOutput = subprocess.check_output(shlex.split(ffprobeCommand))
            logger.success(f"Successfully fetched video properties with ffprobe")
            done = True
        except subprocess.CalledProcessError:
            logger.warning(f"Could not fetch video properties with ffprobe")
            if ffprobeRetries > 0:
                time.sleep(2)
                logger.info(
                    f"Trying {ffprobeRetries} more time(s) to fetch video properties with ffprobe",
                )
                continue
            return None

        ffprobeOutput = ffprobeOutput.decode("utf-8")
        logger.info("-" * 80)
        logger.info("Detecting video properties with ffprobe")
        ffprobeData = json.loads(ffprobeOutput)

        bit_rate = int(int(ffprobeData["format"]["bit_rate"]) / 1000)

        if bit_rate > 0:
            ffprobeData["streams"][0]["bit_rate"] = int(
                int(ffprobeData["format"]["bit_rate"]) / 1000,
            )
        else:
            logger.warning(f"Ignoring estimated bit rate from ffprobe as it is 0.")

        return ffprobeData["streams"][0]

    return None
