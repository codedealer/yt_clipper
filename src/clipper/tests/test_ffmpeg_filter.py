from clipper import ffmpeg_filter

# crop map with redundant 0 duration points at the beginning, middle, and end
CROP_MAP = cropMap = [
    {
        "x": 7.130523,
        "y": 0,
        "crop": "109:426:540:960"
    },
    {
        "x": 7.130523,
        "y": 0,
        "crop": "109:426:540:960"
    },
    {
        "x": 8.08,
        "y": 0,
        "crop": "109:502:540:960"
    },
    {
        "x": 9.59,
        "y": 0,
        "crop": "163:635:450:800"
    },
    {
        "x": 9.59,
        "y": 0,
        "crop": "163:635:450:800"
    },
    {
        "x": 9.59,
        "y": 0,
        "crop": "80:600:470:860"
    },
    {
        "x": 11.22,
        "y": 0,
        "crop": "42:559:513:912"
    },
    {
        "x": 12.827726,
        "y": 0,
        "crop": "42:507:540:960"
    },
    {
        "x": 12.827726,
        "y": 0,
        "crop": "42:507:540:960"
    }
]


def test_getCropFilter(snapshot):
    cropFilter = ffmpeg_filter.getCropFilter("109:426:540:960", CROP_MAP, 60, easeType="easeOutCubic")
    assert cropFilter == snapshot


def test_getZoomPanFilter(snapshot):
    cropFilter, _scale = ffmpeg_filter.getZoomPanFilter(CROP_MAP, 60, easeType="easeOutCubic")
    assert cropFilter == snapshot
