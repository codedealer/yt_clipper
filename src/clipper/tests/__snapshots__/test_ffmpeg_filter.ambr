# serializer version: 1
# name: test_getCropFilter
  "crop='x=(gte(t, 0.0)*lt(t, 0.9494769999999999))*((109)+((109)-(109))*1+((clip(((t-0.0)/0.9494769999999999),0,1))-1)^3)+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*((109)+((163)-(109))*1+((clip(((t-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*((80)+((42)-(80))*1+((clip(((t-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)+between(t, 4.0894770000000005, 5.697203)*((42)+((42)-(42))*1+((clip(((t-4.0894770000000005)/1.6077259999999995),0,1))-1)^3):y=(gte(t, 0.0)*lt(t, 0.9494769999999999))*((426)+((502)-(426))*1+((clip(((t-0.0)/0.9494769999999999),0,1))-1)^3)+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*((502)+((635)-(502))*1+((clip(((t-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*((600)+((559)-(600))*1+((clip(((t-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)+between(t, 4.0894770000000005, 5.697203)*((559)+((507)-(559))*1+((clip(((t-4.0894770000000005)/1.6077259999999995),0,1))-1)^3):w=540:h=960:exact=1'"
# ---
# name: test_getZoomPanFilter
  "crop='x=(gte(t, 0.0)*lt(t, 0.9494769999999999))*max(((1*649.0)+((1*649.0)-(1*649.0))*1+((clip(((t-0.0)/0.9494769999999999),0,1))-1)^3)-1*540, 0)+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*max(((1*649.0)+((1*613.0)-(1*649.0))*1+((clip(((t-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)-1*540, 0)+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*max(((1*550.0)+((1*555.0)-(1*550.0))*1+((clip(((t-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)-1*540, 0)+between(t, 4.0894770000000005, 5.697203)*max(((1*555.0)+((1*582.0)-(1*555.0))*1+((clip(((t-4.0894770000000005)/1.6077259999999995),0,1))-1)^3)-1*540, 0):y=(gte(t, 0.0)*lt(t, 0.9494769999999999))*max(((1*1386.0)+((1*1462.0)-(1*1386.0))*1+((clip(((t-0.0)/0.9494769999999999),0,1))-1)^3)-1*960, 0)+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*max(((1*1462.0)+((1*1435.0)-(1*1462.0))*1+((clip(((t-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)-1*960, 0)+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*max(((1*1460.0)+((1*1471.0)-(1*1460.0))*1+((clip(((t-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)-1*960, 0)+between(t, 4.0894770000000005, 5.697203)*max(((1*1471.0)+((1*1467.0)-(1*1471.0))*1+((clip(((t-4.0894770000000005)/1.6077259999999995),0,1))-1)^3)-1*960, 0):w=1*540:h=1*960:exact=1',scale=w=4*iw:h=4*ih,zoompan=z='((gte(t, 0.0)*lt(t, 0.9494769999999999))*((1.0)+((1.0)-(1.0))*1+((clip(((it-0.0)/0.9494769999999999),0,1))-1)^3)+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*((1.0)+((1.2)-(1.0))*1+((clip(((it-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*((1.148936170212766)+((1.0526315789473684)-(1.148936170212766))*1+((clip(((it-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)+between(t, 4.0894770000000005, 5.697203)*((1.0526315789473684)+((1.0)-(1.0526315789473684))*1+((clip(((it-4.0894770000000005)/1.6077259999999995),0,1))-1)^3))':x='(gte(t, 0.0)*lt(t, 0.9494769999999999))*((((4*109)+((4*109)-(4*109))*1+((clip(((it-0.0)/0.9494769999999999),0,1))-1)^3))-(max(((4*649.0)+((4*649.0)-(4*649.0))*1+((clip(((it-0.0)/0.9494769999999999),0,1))-1)^3)-4*540, 0)))+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*((((4*109)+((4*163)-(4*109))*1+((clip(((it-0.9494769999999999)/1.5099999999999998),0,1))-1)^3))-(max(((4*649.0)+((4*613.0)-(4*649.0))*1+((clip(((it-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)-4*540, 0)))+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*((((4*80)+((4*42)-(4*80))*1+((clip(((it-2.4594769999999997)/1.6300000000000008),0,1))-1)^3))-(max(((4*550.0)+((4*555.0)-(4*550.0))*1+((clip(((it-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)-4*540, 0)))+between(t, 4.0894770000000005, 5.697203)*((((4*42)+((4*42)-(4*42))*1+((clip(((it-4.0894770000000005)/1.6077259999999995),0,1))-1)^3))-(max(((4*555.0)+((4*582.0)-(4*555.0))*1+((clip(((it-4.0894770000000005)/1.6077259999999995),0,1))-1)^3)-4*540, 0)))':y='(gte(t, 0.0)*lt(t, 0.9494769999999999))*((((4*426)+((4*502)-(4*426))*1+((clip(((it-0.0)/0.9494769999999999),0,1))-1)^3))-(max(((4*1386.0)+((4*1462.0)-(4*1386.0))*1+((clip(((it-0.0)/0.9494769999999999),0,1))-1)^3)-4*960, 0)))+(gte(t, 0.9494769999999999)*lt(t, 2.4594769999999997))*((((4*502)+((4*635)-(4*502))*1+((clip(((it-0.9494769999999999)/1.5099999999999998),0,1))-1)^3))-(max(((4*1462.0)+((4*1435.0)-(4*1462.0))*1+((clip(((it-0.9494769999999999)/1.5099999999999998),0,1))-1)^3)-4*960, 0)))+(gte(t, 2.4594769999999997)*lt(t, 4.0894770000000005))*((((4*600)+((4*559)-(4*600))*1+((clip(((it-2.4594769999999997)/1.6300000000000008),0,1))-1)^3))-(max(((4*1460.0)+((4*1471.0)-(4*1460.0))*1+((clip(((it-2.4594769999999997)/1.6300000000000008),0,1))-1)^3)-4*960, 0)))+between(t, 4.0894770000000005, 5.697203)*((((4*559)+((4*507)-(4*559))*1+((clip(((it-4.0894770000000005)/1.6077259999999995),0,1))-1)^3))-(max(((4*1471.0)+((4*1467.0)-(4*1471.0))*1+((clip(((it-4.0894770000000005)/1.6077259999999995),0,1))-1)^3)-4*960, 0)))':d=1:s=540x960:fps=60"
# ---
